### ISSUE TITLE
Error using custom paymaster example

### ISSUE THREAD

**Issue by blenkersA** [matter-labs/zksync-v2-issues#6]
**Describe the issue**
Your Example of Building a custum paymaster isn´t working (https://v2-docs.zksync.io/dev/guide/custom-paymaster-tutorial.html#deploying-the-paymaster).
If I tested it out I came to follwing error if i came to the last script where you use the paymaster:
```
Error: processing response error (body="{\"jsonrpc\":\"2.0\",\"error\":{\"code\":3,\"message\":\"Failed to submit transaction: Failed to validate the transaction. Reason: Validation revert: Paymaster validation error: Failed to transfer funds to the bootloader\",\"data\":{\"code\":104,\"message\":\"Failed to validate the transaction. Reason: Validation revert: Paymaster validation error: Failed to transfer funds to the bootloader\"}},\"id\":58}\n", error={"code":3,"data":{"code":104,"message":"Failed to validate the transaction. Reason: Validation revert: Paymaster validation error: Failed to transfer funds to the bootloader"}}, requestBody="{\"method\":\"eth_sendRawTransaction\",\"params\":[\"0x71f9016d0e8459682f00846553f1008308009294f7a60cce5d4f3abc393951f51d4dff1cfae2212480b84440c10f190000000000000000000000009f03d5226b48267123b8e4d264e619fe2b243ce000000000000000000000000000000000000000000000000000000000000000648201188080820118949f03d5226b48267123b8e4d264e619fe2b243ce083027100c0b84132dfc212c4ef600e27dd1426858c7ff3ec5d857996f8f6576bdb375dc37a232e1e8d9bc48d0de3e5d87ad69f31576df48fd0ee69bcafe524ef531cd652abcb961cf89b94ea7fedec20fd75d8f529448bac114ec23f264ce5b884949431dc000000000000000000000000f7a60cce5d4f3abc393951f51d4dff1cfae22124000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000000\"],\"id\":58,\"jsonrpc\":\"2.0\"}", requestMethod="POST", url="https://zksync2-testnet.zksync.dev", code=SERVER_ERROR, version=web/5.7.1)
    at Logger.makeError (/home/auel/Projekte/daura/api/node_modules/@ethersproject/logger/src.ts/index.ts:269:28)
    at Logger.throwError (/home/auel/Projekte/daura/api/node_modules/@ethersproject/logger/src.ts/index.ts:281:20)
    at /home/auel/Projekte/daura/api/node_modules/@ethersproject/web/src.ts/index.ts:341:28
    at step (/home/auel/Projekte/daura/api/node_modules/@ethersproject/web/lib/index.js:33:23)
    at Object.next (/home/auel/Projekte/daura/api/node_modules/@ethersproject/web/lib/index.js:14:53)
    at fulfilled (/home/auel/Projekte/daura/api/node_modules/@ethersproject/web/lib/index.js:5:58)
    at processTicksAndRejections (node:internal/process/task_queues:95:5) {
  reason: 'processing response error',
  code: 'SERVER_ERROR',
  body: '{"jsonrpc":"2.0","error":{"code":3,"message":"Failed to submit transaction: Failed to validate the transaction. Reason: Validation revert: Paymaster validation error: Failed to transfer funds to the bootloader","data":{"code":104,"message":"Failed to validate the transaction. Reason: Validation revert: Paymaster validation error: Failed to transfer funds to the bootloader"}},"id":58}\n',
  error: Error: Failed to submit transaction: Failed to validate the transaction. Reason: Validation revert: Paymaster validation error: Failed to transfer funds to the bootloader
      at getResult (/home/auel/Projekte/daura/api/node_modules/@ethersproject/providers/src.ts/json-rpc-provider.ts:142:28)
      at processJsonFunc (/home/auel/Projekte/daura/api/node_modules/@ethersproject/web/src.ts/index.ts:383:22)
      at /home/auel/Projekte/daura/api/node_modules/@ethersproject/web/src.ts/index.ts:320:42
      at step (/home/auel/Projekte/daura/api/node_modules/@ethersproject/web/lib/index.js:33:23)
      at Object.next (/home/auel/Projekte/daura/api/node_modules/@ethersproject/web/lib/index.js:14:53)
      at fulfilled (/home/auel/Projekte/daura/api/node_modules/@ethersproject/web/lib/index.js:5:58)
      at processTicksAndRejections (node:internal/process/task_queues:95:5) {
    code: 3,
    data: {
      code: 104,
      message: 'Failed to validate the transaction. Reason: Validation revert: Paymaster validation error: Failed to transfer funds to the bootloader'
    }
  },
  requestBody: '{"method":"eth_sendRawTransaction","params":["0x71f9016d0e8459682f00846553f1008308009294f7a60cce5d4f3abc393951f51d4dff1cfae2212480b84440c10f190000000000000000000000009f03d5226b48267123b8e4d264e619fe2b243ce000000000000000000000000000000000000000000000000000000000000000648201188080820118949f03d5226b48267123b8e4d264e619fe2b243ce083027100c0b84132dfc212c4ef600e27dd1426858c7ff3ec5d857996f8f6576bdb375dc37a232e1e8d9bc48d0de3e5d87ad69f31576df48fd0ee69bcafe524ef531cd652abcb961cf89b94ea7fedec20fd75d8f529448bac114ec23f264ce5b884949431dc000000000000000000000000f7a60cce5d4f3abc393951f51d4dff1cfae22124000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000000"],"id":58,"jsonrpc":"2.0"}',
  requestMethod: 'POST',
  url: 'https://zksync2-testnet.zksync.dev',
  transaction: {
    type: 113,
    nonce: 14,
    maxPriorityFeePerGas: BigNumber { _hex: '0x59682f00', _isBigNumber: true },
    maxFeePerGas: BigNumber { _hex: '0x6553f100', _isBigNumber: true },
    gasLimit: BigNumber { _hex: '0x080092', _isBigNumber: true },
    to: '0xf7A60cce5d4F3abC393951F51D4DFF1cFAe22124',
    value: BigNumber { _hex: '0x00', _isBigNumber: true },
    data: '0x40c10f190000000000000000000000009f03d5226b48267123b8e4d264e619fe2b243ce00000000000000000000000000000000000000000000000000000000000000064',
    chainId: BigNumber { _hex: '0x0118', _isBigNumber: true },
    from: '0x9f03D5226B48267123b8E4d264e619Fe2B243CE0',
    customData: {
      ergsPerPubdata: [BigNumber],
      factoryDeps: [],
      customSignature: '0x32dfc212c4ef600e27dd1426858c7ff3ec5d857996f8f6576bdb375dc37a232e1e8d9bc48d0de3e5d87ad69f31576df48fd0ee69bcafe524ef531cd652abcb961c',
      paymasterParams: [Object]
    },
    hash: '0x7ae627c77ae365eaa264962d8515567383be3901cf9f4f26c780d39c35cc5b50',
    confirmations: 0
  },
  transactionHash: '0x7ae627c77ae365eaa264962d8515567383be3901cf9f4f26c780d39c35cc5b50'
}
```

I used the complete code from the example scripts with just the adjustment of using my account and don´t create a new empty wallet (and comment out the line of code which tests if the wallet has no balnce).

**Expected behavior**
I would expect that the example works without error.

**Used Packages**
"@matterlabs/hardhat-zksync-deploy": "^0.4.0",
"@matterlabs/hardhat-zksync-solc": "^0.3.9",
"@matterlabs/zksync-contracts": "^0.3.3",
"hardhat": "^2.11.2",
"zksync-web3": "^0.9.0",
"ts-node": "^10.9.1",

Hope you can provide some details which is going wrong here

**Comment by bxpana** [comment]
Thanks @blenkersA we'll get looking into this

**Comment by blenkersA** [comment]
Did you guys have more details and maybe an idea what went wrong?
As additional information: I tested it out with the local blockchain [you provide](https://v2-docs.zksync.io/api/hardhat/testing.html) with this it works fine.
Is there a difference between the testnet and and the local blockchain which would explain this issue? I would thought that they are at the same code base

**Comment by bxpana** [comment]
@blenkersA sorry I missed this

It seemed to be that the paymaster didn't have enough to pay the fee based on the script, but we are still doing some tests after regenesis to make sure it works properly.

I didn't know that you were doing local testing before. I'll check on that

**Comment by bxpana** [comment]
@blenkersA It looks like the tutorial was missing some code, but has been resolved and should work. Can you please try again and let us know if it works for you? https://v2-docs.zksync.io/dev/tutorials/custom-aa-tutorial.html

**Comment by blenkersA** [comment]
I will try it out later/tomorrow and give response. Thanks so far

**Comment by blenkersA** [comment]
I could test it today and got another error:
`reason: 'transaction failed',
  code: 'CALL_EXCEPTION',
  transactionHash: '0xcf20469053f0abdaa4d9be2289700790ee14071d269a08a2e2818e7c83f58899',
  transaction: {
    type: 113,
    nonce: 2,
    maxPriorityFeePerGas: BigNumber { _hex: '0x59682f00', _isBigNumber: true },
    maxFeePerGas: BigNumber { _hex: '0x6553f100', _isBigNumber: true },
    gasLimit: BigNumber { _hex: '0x0c913e', _isBigNumber: true },
    to: '0x4ea10ceA482bD0b9CC4352bD34e8dE6474BD356b',
    value: BigNumber { _hex: '0x00', _isBigNumber: true },
    data: '0x40c10f19000000000000000000000000d3323ce9c887ecbd6ec4d77bb36b8cf01e1170160000000000000000000000000000000000000000000000000000000000000064',
    chainId: BigNumber { _hex: '0x0118', _isBigNumber: true },
    from: '0xD3323ce9C887ecBD6eC4d77Bb36b8Cf01E117016',
    customData: {
      ergsPerPubdata: [BigNumber],
      factoryDeps: [],
      customSignature: '0xe03925a703e2cb5677bfc8d345a1b06089ffde412d90d8287c1227b028896bdc3781c730c15e40e1a8c75cc48bc28403d66421c31c9e765c52e0d19fb0d0fbcf1c',
      paymasterParams: [Object]
    },
    hash: '0xcf20469053f0abdaa4d9be2289700790ee14071d269a08a2e2818e7c83f58899',
    confirmations: 0,
    wait: [Function (anonymous)],
    waitFinalize: [AsyncFunction (anonymous)]
  },
  receipt: {
    to: '0x4ea10ceA482bD0b9CC4352bD34e8dE6474BD356b',
    from: '0xD3323ce9C887ecBD6eC4d77Bb36b8Cf01E117016',
    contractAddress: null,
    transactionIndex: 0,
    root: '0x1197951b8e5c5554b19dab2240b32ab87fbbf9ee924e5c5638e4f45769911f19',
    gasUsed: BigNumber { _hex: '0x0caa7b', _isBigNumber: true },
    logsBloom: '0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
    blockHash: '0x1197951b8e5c5554b19dab2240b32ab87fbbf9ee924e5c5638e4f45769911f19',
    transactionHash: '0xcf20469053f0abdaa4d9be2289700790ee14071d269a08a2e2818e7c83f58899',
    logs: [ [Object], [Object], [Object], [Object], [Object] ],
    blockNumber: 116303,
    confirmations: 1,
    cumulativeGasUsed: BigNumber { _hex: '0x00', _isBigNumber: true },
    effectiveGasPrice: BigNumber { _hex: '0x5f5e1000', _isBigNumber: true },
    status: 0,
    type: 113,
    byzantium: true
  }
}`

Whats going on here?

**Comment by blenkersA** [comment]
And yes i tested it with the manual gaslimit and with this you will get the same error as before "Failed to submit transaction: Failed to validate the transaction. Reason: Validation revert: Paymaster validation error: Failed to transfer funds to the bootloader".

I provided some more ether but again this happened and I will not test the exact amount of ether you need on the paymaster to work with it, because the transactions would have fees very less than 0.1 ETH or so.
If this didnßt work we cannot use this because we have more complex transactions we plan to do and we cannot provide millions of eth to the pamaster just that it will work to tranactions with fees of like 0.00001 ETH

**Comment by bxpana** [comment]
@blenkersA it was working a few days ago, but I'll test it again to see if anything has changed. Just to confirm you're running the exact example from the tutorial and it's not working?

**Comment by blenkersA** [comment]
@bxpana Yes I use the custom paymaster example. All my versions are uptodate etc. If it would be an problem with the eth the paymaster holds it would be funny, because if the paymaster has nearly 1 ETH and can not pay for transactions with costs just like 0.0003 ETH :/

**Comment by uF4No** [comment]
Here are a few things that comes to mind:

- Make sure you're using the latest versions of our libraries, specially `"zksync-web3": "^0.11.0"`, `"@matterlabs/zksync-contracts": "^0.4.2"` and `"@matterlabs/hardhat-zksync-deploy": "^0.5.1"`.
- Check your `hardhat.config.ts` and make sure you're using `zksolc` version `1.2.0` and that `compilerSource` is set to `binary`.
- Re-compile and re-deploy your contracts.

I've just ran the code from [the repo](https://github.com/matter-labs/custom-paymaster-tutorial) and managed to run it end to end.

**Comment by blenkersA** [comment]
@uF4No Hi :)
That is exactly what we are using. I just checked out the paymaster repository, installed the dependencies, compiled the contracts and i came to the same problem. I did this twice because i saw that the hardhat config was set to docker so i changed it to binary recompiled and did it again with the same error by using the paymaster.

Error without using the gaslimit:
`Error: processing response error (body="{\"jsonrpc\":\"2.0\",\"error\":{\"code\":3,\"message\":\"Failed to submit transaction: Failed to validate the transaction. Reason: Validation revert: Account validation error: Error function_selector = 0x, data = 0x\",\"data\":{\"code\":104,\"message\":\"Failed to validate the transaction. Reason: Validation revert: Account validation error: Error function_selector = 0x, data = 0x\"}},\"id\":58}\n", error={"code":3,"data":{"code":104,"message":"Failed to validate the transaction. Reason: Validation revert: Account validation error: Error function_selector = 0x, data = 0x"}}, requestBody="{\"method\":\"eth_sendRawTransaction\",\"params\":[\"0x71f9016d808459682f00846553f10083239ce294b59fac50a7c8e49da20a5bc96415c82855ff381580b84440c10f19000000000000000000000000b96142d0988a35f1976058da7c8c6ea13fd9d1200000000000000000000000000000000000000000000000000000000000000064820118808082011894b96142d0988a35f1976058da7c8c6ea13fd9d12083027100c0b841fdb7686fed3c9de95b3fefa1f55ae48945d64ee9b475607ba450df2d876ff974338813c07dc82cdd40a1115306dc4d9fd2770fa25030c71a4cbf55dab2d97adb1cf89b94889e1450d8bcbdcf38c91ef76a343b805d4cc879b884949431dc000000000000000000000000b59fac50a7c8e49da20a5bc96415c82855ff3815000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000000\"],\"id\":58,\"jsonrpc\":\"2.0\"}", requestMethod="POST", url="https://zksync2-testnet.zksync.dev", code=SERVER_ERROR, version=web/5.7.0)
    at Logger.makeError (/home/auel/Projekte/custom-paymaster-tutorial/node_modules/@ethersproject/logger/src.ts/index.ts:269:28)
    at Logger.throwError (/home/auel/Projekte/custom-paymaster-tutorial/node_modules/@ethersproject/logger/src.ts/index.ts:281:20)
    at /home/auel/Projekte/custom-paymaster-tutorial/node_modules/@ethersproject/web/src.ts/index.ts:335:28
    at step (/home/auel/Projekte/custom-paymaster-tutorial/node_modules/@ethersproject/web/lib/index.js:33:23)
    at Object.next (/home/auel/Projekte/custom-paymaster-tutorial/node_modules/@ethersproject/web/lib/index.js:14:53)
    at fulfilled (/home/auel/Projekte/custom-paymaster-tutorial/node_modules/@ethersproject/web/lib/index.js:5:58)
    at processTicksAndRejections (node:internal/process/task_queues:95:5) {
  reason: 'processing response error',
  code: 'SERVER_ERROR',
  body: '{"jsonrpc":"2.0","error":{"code":3,"message":"Failed to submit transaction: Failed to validate the transaction. Reason: Validation revert: Account validation error: Error function_selector = 0x, data = 0x","data":{"code":104,"message":"Failed to validate the transaction. Reason: Validation revert: Account validation error: Error function_selector = 0x, data = 0x"}},"id":58}\n',
  error: Error: Failed to submit transaction: Failed to validate the transaction. Reason: Validation revert: Account validation error: Error function_selector = 0x, data = 0x
      at getResult (/home/auel/Projekte/custom-paymaster-tutorial/node_modules/@ethersproject/providers/src.ts/json-rpc-provider.ts:142:28)
      at processJsonFunc (/home/auel/Projekte/custom-paymaster-tutorial/node_modules/@ethersproject/web/src.ts/index.ts:377:22)
      at /home/auel/Projekte/custom-paymaster-tutorial/node_modules/@ethersproject/web/src.ts/index.ts:314:42
      at step (/home/auel/Projekte/custom-paymaster-tutorial/node_modules/@ethersproject/web/lib/index.js:33:23)
      at Object.next (/home/auel/Projekte/custom-paymaster-tutorial/node_modules/@ethersproject/web/lib/index.js:14:53)
      at fulfilled (/home/auel/Projekte/custom-paymaster-tutorial/node_modules/@ethersproject/web/lib/index.js:5:58)
      at processTicksAndRejections (node:internal/process/task_queues:95:5) {
    code: 3,
    data: {
      code: 104,
      message: 'Failed to validate the transaction. Reason: Validation revert: Account validation error: Error function_selector = 0x, data = 0x'
    }
  },
  requestBody: '{"method":"eth_sendRawTransaction","params":["0x71f9016d808459682f00846553f10083239ce294b59fac50a7c8e49da20a5bc96415c82855ff381580b84440c10f19000000000000000000000000b96142d0988a35f1976058da7c8c6ea13fd9d1200000000000000000000000000000000000000000000000000000000000000064820118808082011894b96142d0988a35f1976058da7c8c6ea13fd9d12083027100c0b841fdb7686fed3c9de95b3fefa1f55ae48945d64ee9b475607ba450df2d876ff974338813c07dc82cdd40a1115306dc4d9fd2770fa25030c71a4cbf55dab2d97adb1cf89b94889e1450d8bcbdcf38c91ef76a343b805d4cc879b884949431dc000000000000000000000000b59fac50a7c8e49da20a5bc96415c82855ff3815000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000000"],"id":58,"jsonrpc":"2.0"}',
  requestMethod: 'POST',
  url: 'https://zksync2-testnet.zksync.dev',
  transaction: {
    type: 113,
    nonce: 0,
    maxPriorityFeePerGas: BigNumber { _hex: '0x59682f00', _isBigNumber: true },
    maxFeePerGas: BigNumber { _hex: '0x6553f100', _isBigNumber: true },
    gasLimit: BigNumber { _hex: '0x239ce2', _isBigNumber: true },
    to: '0xB59faC50A7C8e49Da20A5bC96415C82855ff3815',
    value: BigNumber { _hex: '0x00', _isBigNumber: true },
    data: '0x40c10f19000000000000000000000000b96142d0988a35f1976058da7c8c6ea13fd9d1200000000000000000000000000000000000000000000000000000000000000064',
    chainId: BigNumber { _hex: '0x0118', _isBigNumber: true },
    from: '0xb96142D0988a35F1976058da7C8C6ea13fd9D120',
    customData: {
      ergsPerPubdata: [BigNumber],
      factoryDeps: [],
      customSignature: '0xfdb7686fed3c9de95b3fefa1f55ae48945d64ee9b475607ba450df2d876ff974338813c07dc82cdd40a1115306dc4d9fd2770fa25030c71a4cbf55dab2d97adb1c',
      paymasterParams: [Object]
    },
    hash: '0x90384f06b06919b8dc4e8fc3b9f73a07d33b8f749d4471022bf95cf520f81e58',
    confirmations: 0
  },
  transactionHash: '0x90384f06b06919b8dc4e8fc3b9f73a07d33b8f749d4471022bf95cf520f81e58'
}`

Error with using the gaslimit:
`Error: processing response error (body="{\"jsonrpc\":\"2.0\",\"error\":{\"code\":3,\"message\":\"Failed to submit transaction: Failed to validate the transaction. Reason: Validation revert: Paymaster validation error: Failed to transfer funds to the bootloader\",\"data\":{\"code\":104,\"message\":\"Failed to validate the transaction. Reason: Validation revert: Paymaster validation error: Failed to transfer funds to the bootloader\"}},\"id\":56}\n", error={"code":3,"data":{"code":104,"message":"Failed to validate the transaction. Reason: Validation revert: Paymaster validation error: Failed to transfer funds to the bootloader"}}, requestBody="{\"method\":\"eth_sendRawTransaction\",\"params\":[\"0x71f9016e808459682f00846553f1008401312d0094b59fac50a7c8e49da20a5bc96415c82855ff381580b84440c10f19000000000000000000000000b96142d0988a35f1976058da7c8c6ea13fd9d1200000000000000000000000000000000000000000000000000000000000000064820118808082011894b96142d0988a35f1976058da7c8c6ea13fd9d12083027100c0b8416b3171c987507b347740bafdeaef396524b39f17a84d80de6911bd998a6795923d59135b6a923fee61027ce4623fbf582cbe8d66dc9e5c3047deb13ed481d72c1bf89b94889e1450d8bcbdcf38c91ef76a343b805d4cc879b884949431dc000000000000000000000000b59fac50a7c8e49da20a5bc96415c82855ff3815000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000000\"],\"id\":56,\"jsonrpc\":\"2.0\"}", requestMethod="POST", url="https://zksync2-testnet.zksync.dev", code=SERVER_ERROR, version=web/5.7.0)
    at Logger.makeError (/home/auel/Projekte/custom-paymaster-tutorial/node_modules/@ethersproject/logger/src.ts/index.ts:269:28)
    at Logger.throwError (/home/auel/Projekte/custom-paymaster-tutorial/node_modules/@ethersproject/logger/src.ts/index.ts:281:20)
    at /home/auel/Projekte/custom-paymaster-tutorial/node_modules/@ethersproject/web/src.ts/index.ts:335:28
    at step (/home/auel/Projekte/custom-paymaster-tutorial/node_modules/@ethersproject/web/lib/index.js:33:23)
    at Object.next (/home/auel/Projekte/custom-paymaster-tutorial/node_modules/@ethersproject/web/lib/index.js:14:53)
    at fulfilled (/home/auel/Projekte/custom-paymaster-tutorial/node_modules/@ethersproject/web/lib/index.js:5:58)
    at processTicksAndRejections (node:internal/process/task_queues:95:5) {
  reason: 'processing response error',
  code: 'SERVER_ERROR',
  body: '{"jsonrpc":"2.0","error":{"code":3,"message":"Failed to submit transaction: Failed to validate the transaction. Reason: Validation revert: Paymaster validation error: Failed to transfer funds to the bootloader","data":{"code":104,"message":"Failed to validate the transaction. Reason: Validation revert: Paymaster validation error: Failed to transfer funds to the bootloader"}},"id":56}\n',
  error: Error: Failed to submit transaction: Failed to validate the transaction. Reason: Validation revert: Paymaster validation error: Failed to transfer funds to the bootloader
      at getResult (/home/auel/Projekte/custom-paymaster-tutorial/node_modules/@ethersproject/providers/src.ts/json-rpc-provider.ts:142:28)
      at processJsonFunc (/home/auel/Projekte/custom-paymaster-tutorial/node_modules/@ethersproject/web/src.ts/index.ts:377:22)
      at /home/auel/Projekte/custom-paymaster-tutorial/node_modules/@ethersproject/web/src.ts/index.ts:314:42
      at step (/home/auel/Projekte/custom-paymaster-tutorial/node_modules/@ethersproject/web/lib/index.js:33:23)
      at Object.next (/home/auel/Projekte/custom-paymaster-tutorial/node_modules/@ethersproject/web/lib/index.js:14:53)
      at fulfilled (/home/auel/Projekte/custom-paymaster-tutorial/node_modules/@ethersproject/web/lib/index.js:5:58)
      at processTicksAndRejections (node:internal/process/task_queues:95:5) {
    code: 3,
    data: {
      code: 104,
      message: 'Failed to validate the transaction. Reason: Validation revert: Paymaster validation error: Failed to transfer funds to the bootloader'
    }
  },
  requestBody: '{"method":"eth_sendRawTransaction","params":["0x71f9016e808459682f00846553f1008401312d0094b59fac50a7c8e49da20a5bc96415c82855ff381580b84440c10f19000000000000000000000000b96142d0988a35f1976058da7c8c6ea13fd9d1200000000000000000000000000000000000000000000000000000000000000064820118808082011894b96142d0988a35f1976058da7c8c6ea13fd9d12083027100c0b8416b3171c987507b347740bafdeaef396524b39f17a84d80de6911bd998a6795923d59135b6a923fee61027ce4623fbf582cbe8d66dc9e5c3047deb13ed481d72c1bf89b94889e1450d8bcbdcf38c91ef76a343b805d4cc879b884949431dc000000000000000000000000b59fac50a7c8e49da20a5bc96415c82855ff3815000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000000"],"id":56,"jsonrpc":"2.0"}',
  requestMethod: 'POST',
  url: 'https://zksync2-testnet.zksync.dev',
  transaction: {
    type: 113,
    nonce: 0,
    maxPriorityFeePerGas: BigNumber { _hex: '0x59682f00', _isBigNumber: true },
    maxFeePerGas: BigNumber { _hex: '0x6553f100', _isBigNumber: true },
    gasLimit: BigNumber { _hex: '0x01312d00', _isBigNumber: true },
    to: '0xB59faC50A7C8e49Da20A5bC96415C82855ff3815',
    value: BigNumber { _hex: '0x00', _isBigNumber: true },
    data: '0x40c10f19000000000000000000000000b96142d0988a35f1976058da7c8c6ea13fd9d1200000000000000000000000000000000000000000000000000000000000000064',
    chainId: BigNumber { _hex: '0x0118', _isBigNumber: true },
    from: '0xb96142D0988a35F1976058da7C8C6ea13fd9D120',
    customData: {
      ergsPerPubdata: [BigNumber],
      factoryDeps: [],
      customSignature: '0x6b3171c987507b347740bafdeaef396524b39f17a84d80de6911bd998a6795923d59135b6a923fee61027ce4623fbf582cbe8d66dc9e5c3047deb13ed481d72c1b',
      paymasterParams: [Object]
    },
    hash: '0x267000d0862334bf14d1feac93a015faf9b5c6b109c92409d2ca59a0a28cc45c',
    confirmations: 0
  },
  transactionHash: '0x267000d0862334bf14d1feac93a015faf9b5c6b109c92409d2ca59a0a28cc45c'
}`

**Comment by blenkersA** [comment]
I just checked how the local Blockchain [(local setup](https://github.com/matter-labs/local-setup)) you provide work with this. And it is funny. Here are my foundings:
* If you use your example without adding a manual gaslimit then the example works perfectly -> why behaves the testnet different and throws an error?
* If you use the example with a manual gaslimit (you provide with a comment) it fails with the same error like on testnet
* If you use the example with a manual gaslimit but the paymaster get not only 0.03 ETH but gets 0.3 ETH it worked (but not on the testnet)

What is that different behavior again local Blockchain vs the testnet? For the development it is important to have something which works the same as the "real" network. We can not develop and test our code against the local blockchain just to see that the testnetwork works/behaves different or the other way around. Would be funny if the mainnet will again work different too...

**Comment by uF4No** [comment]
Hi @blenkersA 
We've released a new version of `zksync-web3: v0.11.1` and updated the tutorial to include gas limit calculation for paymaster transactions. Please review [the updated script](https://v2-docs.zksync.io/dev/tutorials/custom-paymaster-tutorial.html#using-the-paymaster) to send the transaction to the paymaster.

Regarding differences between testnet and the local setup, I'll check if there are any changes pending to be released in the docker images used for the local setup.
